# Koneoppiminen: Osa 4: Tukivektorikone

In [ ]:
# Kirjoita tähän tietosi!
student_name = 'Samuli Kotiranta'
student_id = 'AB8349'

## Tehtävä 1
### Aihe: Pulsariaineisto (3 pistettä)

Pulsareita käsittelevä aineisto osoitteesta https://archive.ics.uci.edu/ml/datasets/HTRU2 . Tavoitteena on luokitella aineiston datapisteet joko pulsariksi tai muunlaiseksi kohteeksi. Aineistossa on luokkamuuttuja viimeisessä sarakkeessa. 

1. Lataa aineisto DataFrameen. 
2. Nimeä DataFramen sarakkeet aineiston mukaisesti. Sarakkeiden nimet näet esimerkiksi `.arff`-tiedoston otsakkeesta.  
3. Normalisoi sarakkeiden arvot. 

In [1]:
# Kirjoita toteutuksesi tähän soluun. 

import pandas as pd
from scipy.io.arff import loadarff

# Datan lataaminen
data = 'HTRU_2.csv'
df = pd.read_csv(data, header=None)

# Sarakkeiden nimeäminen
arff = loadarff('HTRU_2.arff')
names_data = pd.DataFrame(arff[0])
df.columns = list(names_data.columns.values)

# Sarakkeiden arvojen normalisointi.
# Tätä ei pidä tehdä ennen datasetin jakamista!
# Tehdään tuossa alempana

In [2]:
# Tehtävän vastaukset. Huom! Älä muokkaa tätä solua, vaan aja se, kun olet suorittanut tehtävän. 
# DataFramen tulisi olla df-muuttujassa. 
print(f'Columns: {df.columns}')
print(f'First row by position: {df.iloc[0]}')

Columns: Index(['Profile_mean', 'Profile_stdev', 'Profile_skewness', 'Profile_kurtosis',
       'DM_mean', 'DM_stdev', 'DM_skewness', 'DM_kurtosis', 'class'],
      dtype='object')
First row by position: Profile_mean        140.562500
Profile_stdev        55.683782
Profile_skewness     -0.234571
Profile_kurtosis     -0.699648
DM_mean               3.199833
DM_stdev             19.110426
DM_skewness           7.975532
DM_kurtosis          74.242225
class                 0.000000
Name: 0, dtype: float64


# oma kommentti
Tässä vastaan tullut `.arff` olikin uusi tuttavuus. Sitä varten ladattiin valmis funktio `scipy`-kirjastosta. Kokeilin aluksi ladata datoja suoraan zipistä, mutta se meni turhan kikkailuksi.

## Tehtävä 2
### Aihe: Koulutus- ja testausaineisto (3 pistettä)

Jaa aineisto koulutus- ja testausaineistoon käyttäen `scikit-learn`-kirjaston [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)-funktiota. Laita puolet aineistosta (_50 %_ ja _50 %_) kumpaankin.

1. Tallenna koulutusaineisto ja sen luokkamuuttuja muuttujiin `X_train`, `y_train`,
2. Tallenna testausaineiston muuttujiin `X_test`, `y_test`.

Käytä satunnaissiemenenä lukua `1492` ja testiaineiston kokoa 50%.

In [3]:
# Kirjoita toteutuksesi tähän soluun. 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

#luokkamuuttujan eristäminen ja aineiston splittaus
X = df.drop(columns='class')
y = df['class']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=1492)

#nyt tehdään skaalaus
sts = StandardScaler().fit(X_train)
X_train = pd.DataFrame(sts.transform(X_train))
X_test = pd.DataFrame(sts.transform(X_test))

In [4]:
# Tehtävän vastaukset. Huom! Älä muokkaa tätä solua, vaan aja se, kun olet suorittanut tehtävän. 
print(f'X_train: {len(X_train)}')
print(f'X_test: {len(X_test)}')
print(f'y_train: {len(y_train)}')
print(f'y_test: {len(y_test)}')

X_train: 8949
X_test: 8949
y_train: 8949
y_test: 8949


# oma kommentti
Tässä on käytössä nyt `StandardScaler` koska tukivekorikone matemaattisena entiteettinä olettaa piirteiden odotusarvon olevan nollassa ja varianssien keskenään samaa luokkaa. Fysiikan osalta voitaisiin lisäksi huomauttaa, että tässä aineistossa yksittäiset mittaukset ovat enemmän tai vähemmän gaussista muotoa. Jakoa ei tehty aikaisemmassa kohdassa, sillä ennen aineiston jakoa kokonaisaineiston informaatiota olisi skaalauksessa välittynyt opetus- ja testijoukkoihin.

## Tehtävä 3
### Aihe: Tukivektorikone (4 pistettä)

1. Käytä `scikit-learn`-kirjaston *Linear Support Vector Classification* -funktiota kouluttaaksesi koulutusaineistollasi luokittelija, joka erottelee pulsarit muista kohteista. 

2. Luo ennusteet testausaineistosi datapisteille `y_pred`-muuttujaan. 

3. Tee `metrics`-moduulilla [luokitteluraportti](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html) testausaineistosi ennusteille. Tallenna raportti `cr`-muuttujaan. Kannattaa myös tulostaa se, jotta näet itse tulokset. 

4. Tee `metrics`-moduulilla **sekaannusmatriisi** (_confusion matrix_) ja tallenna se `cm`-muuttujaan.

In [10]:
# Kirjoita toteutuksesi tähän soluun. 

from sklearn.svm import LinearSVC
from sklearn.metrics import classification_report, confusion_matrix

#tukivektorikone lineaarisella kernelifunktiolla
svm = LinearSVC(max_iter=3000) #iterointikertoen määrä ei saa olla kovin pieni
svm.fit(X_train, y_train)

y_pred = svm.predict(X_test)

#luokitteluraportti ja sekaannusmatriisi
cr = classification_report(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)
#print(cr)

In [11]:
# Tehtävän vastaukset. Huom! Älä muokkaa tätä solua, vaan aja se, kun olet suorittanut tehtävän. 
# Tulokset ovat cm- ja cr-muuttujissa. 
print(f'Classification report:\n {cr}')
print(f'Confusion matrix:\n {cm}')

Classification report:
               precision    recall  f1-score   support

           0       0.98      0.99      0.99      8140
           1       0.94      0.81      0.87       809

    accuracy                           0.98      8949
   macro avg       0.96      0.90      0.93      8949
weighted avg       0.98      0.98      0.98      8949

Confusion matrix:
 [[8097   43]
 [ 156  653]]


# oma kommentti
Tarkkuus jolla lineaarinen tukivektorikone tunnisti pulsarin pulsariksi (0.94) on vielä ihan kohtuullinen. Sensitiivisyydessä (0.81) on sen sijaan toivomisen varaa ja hutejahan tässä tulikin nyt 156 kappaletta, mikä aivan varmasti harmittaisi tutkijaa. Vastaava `F1 score`, tuttavallisemmin Sørensenin–Dicen kerroin, jää melko pieneksi. Tämä on tosin odotettavissakin, koska se ei huomioi oikein menneitä negatiivisia luokitteluja ja negatiivisiahan tässä datassa on kymmenkertainen määrä positiivisiin verrattuna. Negatiivisten osalta luokittelu näyttää (tämän tulkinnassa pitää varoa perustasoharhaa) onnistuneen hyvin, mikä ei myöskään yllätä, koska niitä on niin suuri enemmistö.